File: go/ds/stack/stack.go
package ds

import "fmt"

type Stack struct {
	items []int
	top   int
}

func (stack *Stack) push(item int) {

	if len(stack.items) == 0 {
		stack.top = item
		stack.items = append(stack.items, item)
	}
	stack.items = append(stack.items, item)
}

func (stack *Stack) pop() int {
	if !(len(stack.items) > 0) {
		return 0
	}

	poppedItem := stack.items[len(stack.items)-1]
	stack.items = stack.items[:len(stack.items)-1]

	return poppedItem
}

func (stack *Stack) peek() {
	fmt.Println(stack.items, stack.top)
}

func StackImplementation() {
	stack := Stack{}

	stack.push(1)
	stack.push(2)
	stack.push(3)

	stack.peek()

	stack.pop()
	stack.pop()

	fmt.Println(stack.items)
}

File: java/DataStructures/src/Stack.java
public class Stack {
    private final int maxSize;
    private final long[] stackArray;
    private int top;

    public Stack(int s) {
        maxSize = s;
        stackArray = new long[maxSize];
        top = -1;
    }

    public void push(long j) {
        if (!isFull()) {
            stackArray[++top] = j;
        } else {
            System.out.println("Stack is full, cannot push " + j);
        }
    }

    public long pop() {
        if (!isEmpty()) {
            return stackArray[top--];
        } else {
            System.out.println("Stack is empty, cannot pop");
            return -1; // Return a sentinel value or throw an exception
        }
    }

    public long peek() {
        if (!isEmpty()) {
            return stackArray[top];
        } else {
            System.out.println("Stack is empty, cannot peek");
            return -1; // Return a sentinel value or throw an exception
        }
    }

    public boolean isEmpty() {
        return (top == -1);
    }

    public boolean isFull() {
        return (top == maxSize - 1);
    }
}

class StackImplementation {
    public static void main(String[] args) {
        Stack theStack = new Stack(10);
        theStack.push(20);
        theStack.push(40);
        theStack.push(60);
        theStack.push(80);

        while (!theStack.isEmpty()) {
            long value = theStack.pop();
            System.out.print(value);
            System.out.print(" ");
        }
        System.out.println();
        System.out.println("Stack is empty: " + theStack.isEmpty());
    }
}
